File: D01/csharp/Program.cs
// See https://aka.ms/new-console-template for more information

using System.Diagnostics;


var timestamp = Stopwatch.GetTimestamp();

var numbers = GetNumbers("Input.txt");
var totalDistance = CalculateDistance(numbers.x, numbers.y);
var similarityScore = CalculateSimilarityScore(numbers.x, numbers.y);

var elapsedMilliseconds = Stopwatch.GetElapsedTime(timestamp).TotalMilliseconds;

Console.WriteLine($"The total distance is: {totalDistance}");
Console.WriteLine($"The similarity score is: {similarityScore}");
Console.WriteLine($"Elapsed time: {elapsedMilliseconds}ms");

await Task.Delay(1000);

var timestampLinq = Stopwatch.GetTimestamp();

var numbersLinq = GetNumbersLinq("Input.txt");
var totalDistanceLinq = CalculateDistanceLinq(numbersLinq.x, numbersLinq.y);
var similarityScoreLinq = CalculateSimilarityScoreLinq(numbersLinq.x, numbersLinq.y);

var elapsedMillisecondsLinq = Stopwatch.GetElapsedTime(timestampLinq).TotalMilliseconds;

Console.WriteLine($"The total distance is linq: {totalDistanceLinq}");
Console.WriteLine($"The similarity score is linq: {similarityScoreLinq}");
Console.WriteLine($"Elapsed time linq: {elapsedMillisecondsLinq}ms");

return;


(int[] x, int[] y) GetNumbers(in string filePath)
{
    var lines = File.ReadLines(filePath).ToArray();

    var x = new int[lines.Length];
    var y = new int[lines.Length];

    for (int i = 0; i < lines.Length; i++)
    {
        var parts = lines[i].Split("   ");
        x[i] = int.Parse(parts[0]);
        y[i] = int.Parse(parts[1]);
    }

    Array.Sort(x);
    Array.Sort(y);

    return (x, y);
}


(IEnumerable<int> x, IEnumerable<int> y) GetNumbersLinq(in string filePath)
{
    var linq = File
        .ReadLines(filePath)
        .Select(line => line.Split("   "))
        .Select(parts => (int.Parse(parts[0]), int.Parse(parts[1])))
        .ToArray();
    return (linq.Select(x => x.Item1).Order(), linq.Select(x => x.Item2).Order());
}

int CalculateDistance(int[] a, int[] b)
{
    var sum = 0;
    for (int i = 0; i < Math.Min(a.Length, b.Length); i++) sum += Math.Abs(a[i] - b[i]);
    return sum;
}


int CalculateDistanceLinq(IEnumerable<int> a, IEnumerable<int> b)
    => a.Zip(b, (x, y) => Math.Abs(x - y)).Sum();

int CalculateSimilarityScore(int[] a, int[] b)
{
    var occurences = new Dictionary<int, int>(b.Length);
    foreach (var n in b)
    {
        if (occurences.TryGetValue(n, out var value))
            occurences[n] = ++value;
        else
            occurences[n] = 1;
    }

    var sum = 0;
    foreach (var n in a)
    {
        if (occurences.TryGetValue(n, out var value)) sum += n * value;
    }

    return sum;
}
    
int CalculateSimilarityScoreLinq(IEnumerable<int> a, IEnumerable<int> b)
{
    var occurences = b
        .GroupBy(x => x)
        .ToDictionary(x => x.Key, x => x.Count());
    return a.Sum(x => occurences.TryGetValue(x, out var value) ? x * value : 0);
}

File: D01/fsharp/Program.fs
// For more information see https://aka.ms/fsharp-console-apps

open System
open System.Diagnostics
open System.IO
open System.Collections.Generic

let readFile (filePath: string) =
    // open file and read all lines
    let lines = File.ReadLines(filePath) |> Seq.toArray 
    
    let x: int[] = Array.zeroCreate lines.Length
    let y: int[] = Array.zeroCreate lines.Length
    
    // load values
    lines
    |> Array.iteri (fun index line ->
        let parts = line.Split("   ");
        x[index] <- int parts[0]
        y[index] <- int parts[1]
    )
    
    Array.sortInPlace x
    Array.sortInPlace y
    
    // return the two queues
    x, y
    
let addOrIncrement key (dict: Dictionary<int, int>) =
    match dict.TryGetValue(key) with
    | true, value -> dict[key] <- value + 1
    | false, _ -> dict.Add(key, 1)
    
let rec calculateDistance index (a: int[]) (b: int[]) =
    match index with
    | _ when index >= min a.Length b.Length -> 0
    | _ -> abs(a[index] - b[index]) + calculateDistance (index + 1) a b
    
let calculateSimilarityScore (a: int[]) (b: int[]) =
    let occurenceMap = Dictionary<int, int>(b.Length)
    Array.iter (fun key -> addOrIncrement key occurenceMap) b
    Array.fold (fun acc key -> acc + key * (match occurenceMap.TryGetValue(key) with
                                           | true, value -> value
                                           | _ -> 0)) 0 a
    
let timestamp = Stopwatch.GetTimestamp()
let leftInput, rightInput = readFile "Input.txt"
let totalDistance = calculateDistance 0 leftInput rightInput
let similarityScore = calculateSimilarityScore leftInput rightInput
let elapsedMilliseconds = Stopwatch.GetElapsedTime(timestamp).TotalMilliseconds

printfn $"The total distance is: {totalDistance}"
printfn $"The similarity score is: {similarityScore}"
printfn $"Elapsed time: {elapsedMilliseconds}ms"

